tools/clang/lib/SPIRV/RemoveBufferBlockVisitor.cpp

```diff
@@ -71,15 +72,88 @@ bool RemoveBufferBlockVisitor::visitInstruction(SpirvInstruction *inst) {
 
   // For all instructions, if the result type is a pointer pointing to a struct
   // with StorageBuffer interface, the storage class must be updated.
-  if (auto *ptrResultType = dyn_cast<SpirvPointerType>(inst->getResultType())) {
-    if (hasStorageBufferInterfaceType(ptrResultType->getPointeeType()) &&
-        ptrResultType->getStorageClass() != spv::StorageClass::StorageBuffer) {
-      inst->setStorageClass(spv::StorageClass::StorageBuffer);
-      inst->setResultType(context.getPointerType(
-          ptrResultType->getPointeeType(), spv::StorageClass::StorageBuffer));
+  const auto *instType = inst->getResultType();
+  const auto *newInstType = instType;
+  spv::StorageClass newInstStorageClass = spv::StorageClass::Max;
+  if (updateStorageClass(instType, &newInstType, &newInstStorageClass)) {
+    inst->setResultType(newInstType);
+    inst->setStorageClass(newInstStorageClass);
+  }
+
+  return true;
+}
+
+bool RemoveBufferBlockVisitor::updateStorageClass(
+    const SpirvType *type, const SpirvType **newType,
+    spv::StorageClass *newStorageClass) {
+  if (!isa<SpirvPointerType>(type))
+    return false;
+
+  auto *ptrType = cast<SpirvPointerType>(type);
```

Comment:
Can we change the above code
```
  if (!isa<SpirvPointerType>(type))
    return false;

  auto *ptrType = cast<SpirvPointerType>(type);
```
to
```
auto *ptrType = dyn_cast<SpirvPointerType>(type);
if (ptrType == nullptr)
  return false;
```
?

I think the performance might be almost similar, but `cast` looks unsafe for me. I mean your code is ok but later someone can update the code in a wrong way.

Reply:
I agree, this is a better way. Done.